Source/Test/Rewriting/Passes/AssemblyRewriter.cs

```diff
@@ -158,16 +158,48 @@ protected virtual Instruction VisitInstruction(Instruction instruction)
         protected MethodReference RewriteMethodReference(MethodReference method, ModuleDefinition module, string matchName = null)
         {
             MethodReference result = method;
-
+            MethodDefinition resolvedMethod = method.Resolve();
             TypeReference declaringType = this.RewriteDeclaringTypeReference(method);
-            if (method.DeclaringType == declaringType ||
-                !this.TryResolve(method, out MethodDefinition resolvedMethod))
+            var resolvedType = Resolve(declaringType);
+
+            if (resolvedMethod == null)
```

Comment:
Is the logic in this `if` branch that returns a previously rewritten method an optimization, or is this fundamental to your changes/fixes?

The reason I am asking is because I am wondering if it is safe to return an already rewritten method and reuse it? I remember some very weird mutability issues that we faced ages ago, when we were trying to cache rewritten methods or types, and then fields of them were modified in later passes, and ended up always reconstructing types/methods, but I cannot recall the details, besides that it had bite us :)

Reply:
This was complex.  The problem with the previous code is that it was assuming too early that the methodReference does not need to be re-written if it had the same DeclaringType, but that is not true, because the MethodReferenced may have a ReturnType or ParameterType that needs to be rewritten.

Reply:
This was complex. The problem with the previous code is that it was assuming too early that the methodReference does not need to be re-written if it had the same DeclaringType, but that is not true, because the MethodReferenced may have a ReturnType or ParameterType that needs to be rewritten.